File: src/main/java/gg/mineral/practice/match/data/MatchStatisticCollector.java
package gg.mineral.practice.match.data;

import java.util.List;

import org.bukkit.ChatColor;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;

import gg.mineral.api.collection.GlueList;
import gg.mineral.practice.entity.Profile;
import gg.mineral.practice.util.math.MathUtil;
import gg.mineral.practice.util.messages.StringUtil;
import lombok.Getter;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class MatchStatisticCollector {
    final Profile profile;
    @Getter
    int hitCount, currentCombo, longestCombo, averageCombo, highestCps, wTapCount, wTapAccuracy,
            potionsThrown, potionsMissed, potionsStolen, potionsRemaining, potionAccuracy, soupsRemaining,
            remainingHealth;
    int clicks;
    long clickCounterStart;
    @Getter
    ItemStack[] inventoryContents;
    @Getter
    ItemStack helmet, chestplate, leggings, boots;
    @Getter
    List<String> potionEffectStrings;
    boolean active = false;
    @Getter
    boolean alive = false;

    public void start() {
        if (active)
            throw new IllegalStateException("Already started");

        active = true;
        currentCombo = hitCount = longestCombo = averageCombo = highestCps = wTapCount = wTapAccuracy = potionsThrown = potionsMissed = potionsStolen = potionsRemaining = potionAccuracy = soupsRemaining = clicks = 0;
        clickCounterStart = System.currentTimeMillis();
        potionEffectStrings = new GlueList<String>();
    }

    public void end(boolean alive) {
        if (!active)
            throw new IllegalStateException("Not been started yet.");

        active = false;
        this.alive = alive;
        this.potionsRemaining = profile.getInventory().getNumber(Material.POTION, (short) 16421);
        this.soupsRemaining = profile.getInventory().getNumber(Material.MUSHROOM_SOUP);
        this.remainingHealth = profile.getPlayer().isDead() ? 0 : (int) profile.getPlayer().getHealth();
        this.potionAccuracy = (int) (100 - (getPotionsMissed() * 100D
                / getPotionsThrown()));
        this.wTapAccuracy = (int) (getWTapCount() * 100D
                / getHitCount());
        this.inventoryContents = profile.getInventory().getContents();
        this.helmet = profile.getInventory().getHelmet();
        this.chestplate = profile.getInventory().getChestplate();
        this.leggings = profile.getInventory().getLeggings();
        this.boots = profile.getInventory().getBoots();

        for (final PotionEffect potionEffect : profile.getPlayer().getActivePotionEffects()) {
            final String romanNumeral = MathUtil.convertToRomanNumeral(potionEffect.getAmplifier() + 1);
            final String effectName = StringUtil.toNiceString(potionEffect.getType().getName().toLowerCase());
            final String duration = MathUtil.convertTicksToMinutes(potionEffect.getDuration());
            potionEffectStrings.add(ChatColor.YELLOW.toString() + ChatColor.BOLD + "* " + ChatColor.WHITE + effectName
                    + " " + romanNumeral + ChatColor.GRAY + " (" + duration + ")");
        }
    }

    public void increaseHitCount() {
        if (!active)
            return;

        hitCount++;
        currentCombo++;

        if (profile.getPlayer().getHandle().isSprinting())
            wTapCount++;

        if (currentCombo > 1) {
            averageCombo += currentCombo;
            averageCombo /= 2;
        }

        longestCombo = Math.max(currentCombo, longestCombo);
    }

    public void resetCombo() {
        if (!active)
            return;

        currentCombo = 0;
    }

    public void clearHitCount() {
        if (!active)
            return;

        hitCount = 0;
    }

    public void thrownPotion(boolean missed) {
        if (!active)
            return;

        potionsThrown++;

        if (missed)
            potionsMissed++;
    }

    public void stolenPotion() {
        if (!active)
            return;

        potionsStolen++;
    }

    public void click() {
        if (!active)
            return;

        if (System.currentTimeMillis() - clickCounterStart > 1000) {
            clickCounterStart = System.currentTimeMillis();
            highestCps = Math.max(clicks, highestCps);
            clicks = 1;
        }

        clicks++;
    }

    public String[] getPotionEffectStringArray() {
        return potionEffectStrings.toArray(new String[0]);
    }

}

File: src/main/kotlin/gg/mineral/practice/match/data/MatchStatisticCollector.kt
package gg.mineral.practice.match.data

import gg.mineral.api.collection.GlueList
import gg.mineral.practice.entity.Profile
import gg.mineral.practice.util.math.MathUtil
import gg.mineral.practice.util.messages.StringUtil
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import java.util.*
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sqrt

class MatchStatisticCollector(val profile: Profile) {
    var hitCount: Int = 0
    private var currentCombo: Int = 0
    var longestCombo: Int = 0
    var averageCombo: Int = 0
    var highestCps: Int = 0
    private var wTapCount: Int = 0
    var wTapAccuracy: Int = 0
    var potionsThrown: Int = 0
    var potionsMissed: Int = 0
    var potionsStolen: Int = 0
    var potionsRemaining: Int = 0
    var potionAccuracy: Int = 0
    var soupsRemaining: Int = 0
    var remainingHealth: Int = 0
    private var clicks: Int = 0
    var totalClicks = 0
    private var clickCounterStart: Long = 0
    lateinit var inventoryContents: Array<ItemStack?>
    var helmet: ItemStack? = null
    var chestplate: ItemStack? = null
    var leggings: ItemStack? = null
    var boots: ItemStack? = null
    private var potionEffectStrings: MutableList<String>? = null
    private var active: Boolean = false
    var alive: Boolean = false
    val potionEffectStringArray: Array<String>
        get() = potionEffectStrings!!.toTypedArray<String>()
    private val cpsRecord = mutableListOf<Int>()
    val averageCps: Int
        get() = if (cpsRecord.isEmpty()) 0 else cpsRecord.average().roundToInt()
    val cpsDeviation: Int
        get() {
            if (cpsRecord.isEmpty()) return 0
            val mean = averageCps.toDouble()
            val variance = cpsRecord.map { (it - mean) * (it - mean) }.average()
            return sqrt(variance).roundToInt()
        }

    fun start() {
        check(!active) { "Already started" }

        active = true
        clicks = 0
        soupsRemaining = 0
        potionAccuracy = 0
        potionsRemaining = 0
        potionsStolen = 0
        potionsMissed = 0
        potionsThrown = 0
        wTapAccuracy = 0
        wTapCount = 0
        highestCps = 0
        averageCombo = 0
        longestCombo = 0
        remainingHealth = 0
        hitCount = 0
        currentCombo = 0
        clickCounterStart = profile.clientTimeMillis
        potionEffectStrings = GlueList()
    }

    fun end(alive: Boolean) {
        if (!active) return

        active = false
        this.alive = alive
        this.potionsRemaining = profile.inventory.getNumber(Material.POTION, 16421.toShort())
        this.soupsRemaining = profile.inventory.getNumber(Material.MUSHROOM_SOUP)
        this.remainingHealth = profile.player?.let { if (it.isDead) 0 else it.health.toInt() } ?: 0
        this.potionAccuracy = (100 - (potionsMissed * 100.0
                / potionsThrown)).toInt()
        this.wTapAccuracy = (wTapCount * 100.0
                / hitCount).toInt()
        this.inventoryContents = profile.inventory.contents
        this.helmet = profile.inventory.helmet
        this.chestplate = profile.inventory.chestplate
        this.leggings = profile.inventory.leggings
        this.boots = profile.inventory.boots

        profile.player?.apply {
            for (potionEffect in activePotionEffects) {
                val romanNumeral = MathUtil.convertToRomanNumeral(potionEffect.amplifier + 1)
                val effectName = StringUtil.toNiceString(potionEffect.type.name.lowercase(Locale.getDefault()))
                val duration = MathUtil.convertTicksToMinutes(potionEffect.duration)
                potionEffectStrings!!.add(
                    (ChatColor.YELLOW.toString() + ChatColor.BOLD + "* " + ChatColor.WHITE + effectName
                            + " " + romanNumeral + ChatColor.GRAY + " (" + duration + ")")
                )
            }
        }
    }

    fun increaseHitCount() {
        if (!active) return

        hitCount++
        currentCombo++

        if (profile.player?.handle?.isSprinting == true) wTapCount++

        if (currentCombo > 1) {
            averageCombo += currentCombo
            averageCombo /= 2
        }

        longestCombo = max(currentCombo.toDouble(), longestCombo.toDouble()).toInt()
    }

    fun resetCombo() {
        if (!active) return

        currentCombo = 0
    }

    fun clearHitCount() {
        if (!active) return

        hitCount = 0
    }

    fun thrownPotion(missed: Boolean) {
        if (!active) return

        potionsThrown++

        if (missed) potionsMissed++
    }

    fun stolenPotion() {
        if (!active) return

        potionsStolen++
    }

    fun click() {
        if (!active) return

        if (profile.clientTimeMillis - clickCounterStart >= 1000) {
            clickCounterStart = profile.clientTimeMillis
            highestCps = max(clicks.toDouble(), highestCps.toDouble()).toInt()
            if (clicks > min(1, highestCps)) cpsRecord.add(clicks)
            clicks = 0
        }

        clicks++
        totalClicks++
    }
}
